src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt

```diff
@@ -38,15 +40,16 @@ class RsExtractTraitHandler : RefactoringActionHandler {
     override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
         val offset = editor.caretModel.offset
         val element = file.findElementAt(offset) ?: return
-        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
-        if (impl.traitRef != null) return
-        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+        val traitOrImpl = element.ancestorOrSelf<RsTraitOrImpl>() ?: return
+        if (traitOrImpl is RsImplItem && traitOrImpl.traitRef != null) return
```

Comment:
[minor] `if` checks could be replaced with `when` to emphasize that it is either `RsImplItem` or `RsTraitItem`, like `org.rust.lang.core.types.ExtensionsKt#getSelfType` for example

Reply:
We have second condition and `return` here, so I am not sure how best to do it